crates/oxc_codegen/src/code_buffer.rs

```diff
@@ -0,0 +1,422 @@
+    /// let source_text = code.take_source_text();
+    /// ```
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create a new, empty `CodeBuffer` with the specified capacity.
+    ///
+    /// The buffer will be able to hold at least `capacity` bytes without
+    /// reallocating. This method is allowed to allocate for more bytes than
+    /// `capacity`. If `capacity` is 0, the buffer will not allocate.
+    ///
+    /// It is important to note that although the returned buffer has the
+    /// minimum *capacity* specified, the buffer will have a zero *length*.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new capacity exceeds `isize::MAX` _bytes_.
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self { buf: Vec::with_capacity(capacity) }
+    }
+
+    /// Returns the number of bytes in this buffer.
+    ///
+    /// This is _not_ the same as the number of characters in the buffer, since
+    /// non-ASCII characters require multiple bytes.
+    pub fn len(&self) -> usize {
+        self.buf.len()
+    }
+
+    /// Returns `true` if this buffer contains no characters.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
+    /// let mut code = CodeBuffer::new();
+    /// assert!(code.is_empty());
+    ///
+    /// code.push_char('c');
+    /// assert!(!code.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.buf.is_empty()
+    }
+
+    /// Reserves capacity for at least `additional` more characters in the given
+    /// `CodeBuffer`. The buffer may reserve more space to speculatively avoid
+    /// frequent reallocations. After calling `reserve`, capacity will be
+    /// greater than or equal to `self.len() + additional`. Does nothing if
+    /// capacity is already sufficient.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new capacity exceeds `isize::MAX` _bytes_.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut code = CodeBuffer::default();
+    /// code.reserve(10);
+    /// ```
+    #[inline]
+    pub fn reserve(&mut self, additional: usize) {
+        self.buf.reserve(additional);
+    }
+
+    /// Peek the `n`th character from the end of the buffer.
+    /// When `n` is zero, the last character is returned. Returns [`None`] if
+    /// `n` exceeds the length of the buffer.
+    ///
+    /// ## Examples
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
+    /// let mut code = CodeBuffer::new();
+    /// code.print_str("foo");
+    ///
+    /// assert_eq!(code.peek_nth(0), Some('o'));
+    /// assert_eq!(code.peek_nth(2), Some('f'));
+    /// assert_eq!(code.peek_nth(3), None);
+    /// ```
+    #[inline]
+    #[must_use = "Peeking is pointless if the peeked char isn't used"]
+    pub fn peek_nth(&self, n: usize) -> Option<char> {
+        // SAFETY: `buf` is a valid UTF-8 string because of invariants upheld by
+        // CodeBuffer
```

Comment:
```suggestion
        // SAFETY: `buf` is a valid UTF-8 string because of invariants upheld by CodeBuffer
```
It looks like better to keep one line